numpyro/infer/renyi_elbo.py

```diff
@@ -58,64 +56,14 @@ def single_particle_elbo(rng, stop_gradient=False):
-        return np.mean(vmap(single_particle_renyi_elbo)(rng_keys))
-
-    def loss_and_grads(self, rng, param_map, model, guide, *args, **kwargs):
-        r"""
-        :returns: returns an estimate of the ELBO
-        :rtype: float
-        Computes the ELBO as well as the surrogate ELBO that is used to form the gradient estimator.
-        Performs backward on the latter. Num_particle many samples are used to form the estimators.
-        """
-        def single_particle_elbo(rng, stop_gradient=False):
-            model_seed, guide_seed = random.split(rng)
-            seeded_model = seed(model, model_seed)
-            seeded_guide = seed(guide, guide_seed)
-            guide_log_density, guide_trace = log_density(seeded_guide, args, kwargs, param_map)
-            # NB: we only want to substitute params not available in guide_trace
-            model_param_map = {k: v for k, v in param_map.items() if k not in guide_trace}
-            seeded_model = replay(seeded_model, guide_trace)
-            model_log_density, _ = log_density(seeded_model, args, kwargs, model_param_map)
-
-            # log p(z) - log q(z)
-            if stop_gradient:
-                elbo = stop_gradient(model_log_density) - stop_gradient(guide_log_density)
-            else:
-                elbo = model_log_density - guide_log_density
-            # Return (-elbo) since by convention we do gradient descent on a loss and
-            # the ELBO is a lower bound that needs to be maximized.
-            return -elbo
-
-        def single_particle_renyi_elbo(rng):
-            elbo = - single_particle_elbo(rng)
-            # not sure if this is numerically stable
-            renyi_elbo = np.exp((1. - self.alpha) * elbo) / (1. - self.alpha)
-            return -renyi_elbo
-
+            
         rng_keys = random.split(rng, self.num_particles)
-        stop_gradients = [True for i in range(self.num_particles)]
-        # not sure how to pass multiple args into vmap
-        elbo_particles = vmap(single_particle_elbo)(rng_keys, stop_gradients)
-        surrogate_elbo_particles = vmap(single_particle_renyi_elbo)(rng_keys)
-        log_weights = (1. - self.alpha) * elbo_particles
-        log_mean_weight = logsumexp(log_weights, axis=0) - np.log(self.num_particles)
-        elbo = np.sum(log_mean_weight) / (1. - self.alpha)
-        normalized_weights = np.exp(log_weights - log_mean_weight)
-        surrogate_elbo = np.sum(normalized_weights * surrogate_elbo_particles) / self.num_particles
-        surrogate_loss = -surrogate_elbo #what to do with this?
-        return -elbo
+        elbos = vmap(single_particle_elbo)(rng_keys)
+        scaled_elbos = (1. - self.alpha) * elbos
+        weights = scaled_elbos - logsumexp(scaled_elbos)
+        renyi_elbo = np.mean(np.exp(logsumexp(scaled_elbos))) / (1. - self.alpha)
+        return stop_gradient(renyi_elbo - weights.dot(elbos)) + stop_gradient(weights).dot(elbos)
```

Comment:
I used `.dot` just for explanation. Probably you will want to use `np.dot(weights, elbos)`

Reply:
Sorry, the fastest way is probably (not important because these computations are cheap comparing to elbo computation)
```
weighted_elbo = stop_gradient(weights).dot(elbos)
return stop_gradient(renyi_elbo - weighted_elbo) + weighted_elbo
```

Reply:
@TuanNguyen27 the above version is faster and seems more reasonable. Could you double-check that we need to take the mean over num_particles for `weighted_elbo` and return the negative sign of `differentiable_renyi_elbo`?

Reply:
I meant divided by `self.num_particles` for the term `np.dot(stop_gradient(weights), elbos)`, so that the gradient is the average over num_particles.